Ycheck failure in covariant java varargs

This involves Scala code calling a Java varargs methods.

Java
```
  class Animal {}
  class Dog extends Animal {}
  class J {
    void foo(Animal... animal) {}
  }
```

Scala
```
class S {
  val j = new J()
  val x: Array[Dog] = ???
  j.foo(x: _*)
}
```

Compiling without -Ycheck succeeds, but -Ycheck:refchecks fails:
```
exception while typing this.j.foo(this.x) of class class dotty.tools.dotc.ast.Trees$Apply # 802
exception while typing @scala.annotation.internal.SourceFile("tests/neg/java-varargs-cov/S.scala") 
  class
 S() extends Object() { 
  val j: J = new J()
  val x: Array[Dog] = ???
  this.j.foo(this.x)
} of class class dotty.tools.dotc.ast.Trees$TypeDef # 804
exception while typing package <empty> {
  @scala.annotation.internal.SourceFile("tests/neg/java-varargs-cov/S.scala") 
    class
   S() extends Object() { 
    val j: J = new J()
    val x: Array[Dog] = ???
    this.j.foo(this.x)
  }
} of class class dotty.tools.dotc.ast.Trees$PackageDef # 805
*** error while checking tests/neg/java-varargs-cov/S.scala after phase refchecks ***
exception occurred while compiling tests/neg/java-varargs-cov/S.scala
Exception in thread "main" java.lang.AssertionError: assertion failed: found:    Array[Dog](S.this.x)
required: Array[Animal]


tree = this.x
	at scala.Predef$.assert(Predef.scala:219)
	at dotty.tools.dotc.transform.TreeChecker$Checker.adapt(TreeChecker.scala:460)
	at dotty.tools.dotc.typer.ProtoTypes$FunProto.typedArg(ProtoTypes.scala:302)
	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.typedArg(Applications.scala:663)
	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.typedArg(Applications.scala:661)
	at dotty.tools.dotc.typer.Applications$Application.addTyped$1(Applications.scala:434)
	at dotty.tools.dotc.typer.Applications$Application.matchArgs(Applications.scala:472)
	at dotty.tools.dotc.typer.Applications$Application.init(Applications.scala:255)
	at dotty.tools.dotc.typer.Applications$TypedApply.<init>(Applications.scala:564)
	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.<init>(Applications.scala:662)
	at dotty.tools.dotc.typer.Applications.simpleApply$1(Applications.scala:722)
	at dotty.tools.dotc.typer.Applications.$anonfun$typedApply$7(Applications.scala:747)
	at dotty.tools.dotc.typer.Typer.tryEither(Typer.scala:1983)
	at dotty.tools.dotc.typer.Applications.$anonfun$typedApply$1(Applications.scala:748)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:37)
	at dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:692)
	at dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:793)
	at dotty.tools.dotc.typer.Applications.typedApply$(Applications.scala:690)
	at dotty.tools.dotc.typer.Typer.typedApply(Typer.scala:82)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1817)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1868)
	at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:112)
	at dotty.tools.dotc.transform.TreeChecker$Checker.typedUnadapted(TreeChecker.scala:282)
	at dotty.tools.dotc.typer.Typer.$anonfun$typed$2(Typer.scala:1899)
	at dotty.tools.dotc.reporting.trace$.apply(trace.scala:40)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1895)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1911)
	at dotty.tools.dotc.transform.TreeChecker$Checker.typed(TreeChecker.scala:270)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1951)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1964)
	at dotty.tools.dotc.transform.TreeChecker$Checker.typedStats(TreeChecker.scala:446)
	at dotty.tools.dotc.typer.Typer.$anonfun$typedClassDef$1(Typer.scala:1564)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:37)
	at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:1487)
	at dotty.tools.dotc.transform.TreeChecker$Checker.typedClassDef(TreeChecker.scala:399)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1808)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1867)
	at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:112)
	at dotty.tools.dotc.transform.TreeChecker$Checker.typedUnadapted(TreeChecker.scala:282)
	at dotty.tools.dotc.typer.Typer.$anonfun$typed$2(Typer.scala:1899)
	at dotty.tools.dotc.reporting.trace$.apply(trace.scala:40)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1895)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1911)
	at dotty.tools.dotc.transform.TreeChecker$Checker.typed(TreeChecker.scala:270)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1930)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1964)
	at dotty.tools.dotc.transform.TreeChecker$Checker.typedStats(TreeChecker.scala:446)
	at dotty.tools.dotc.typer.Typer.$anonfun$typedPackageDef$1(Typer.scala:1674)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:37)
	at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:1667)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1847)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1868)
	at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:112)
	at dotty.tools.dotc.transform.TreeChecker$Checker.typedUnadapted(TreeChecker.scala:282)
	at dotty.tools.dotc.typer.Typer.$anonfun$typed$2(Typer.scala:1899)
	at dotty.tools.dotc.reporting.trace$.apply(trace.scala:40)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1895)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1911)
	at dotty.tools.dotc.transform.TreeChecker$Checker.typed(TreeChecker.scala:270)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1975)
	at dotty.tools.dotc.transform.TreeChecker.check(TreeChecker.scala:132)
	at dotty.tools.dotc.transform.TreeChecker.run(TreeChecker.scala:104)
	at dotty.tools.dotc.core.Phases$Phase.$anonfun$runOn$1(Phases.scala:298)
	at scala.collection.immutable.List.map(List.scala:283)
	at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:296)
	at dotty.tools.dotc.core.Phases$Phase.runOn$(Phases.scala:295)
	at dotty.tools.dotc.transform.TreeChecker.runOn(TreeChecker.scala:48)
	at dotty.tools.dotc.Run.$anonfun$compileUnits$3(Run.scala:174)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.util.Stats$.trackTime(Stats.scala:49)
	at dotty.tools.dotc.Run.$anonfun$compileUnits$2(Run.scala:171)
	at dotty.tools.dotc.Run.$anonfun$compileUnits$2$adapted(Run.scala:169)
	at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
	at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:194)
	at dotty.tools.dotc.Run.runPhases$1(Run.scala:169)
	at dotty.tools.dotc.Run.$anonfun$compileUnits$1(Run.scala:194)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:90)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:149)
	at dotty.tools.dotc.Run.compileSources(Run.scala:136)
	at dotty.tools.dotc.Run.compile(Run.scala:120)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:31)
	at dotty.tools.dotc.Driver.process(Driver.scala:134)
	at dotty.tools.dotc.Driver.process(Driver.scala:103)
	at dotty.tools.dotc.Driver.process(Driver.scala:115)
	at dotty.tools.dotc.Driver.main(Driver.scala:142)

```

--- a/compiler/src/dotty/tools/dotc/core/TypeApplications.scala
+++ b/compiler/src/dotty/tools/dotc/core/TypeApplications.scala
@@ -435,15 +435,19 @@ class TypeApplications(val self: Type) extends AnyVal {
     case _ => if (self.isMatch) MatchAlias(self) else TypeAlias(self)
   }
 
-  /** Translate a type of the form From[T] to To[T], keep other types as they are.
+  /** Translate a type of the form From[T] to either To[T] or To[_ <: T] (if `wildcardArg` is set). Keep other types as they are.
    *  `from` and `to` must be static classes, both with one type parameter, and the same variance.
    *  Do the same for by name types => From[T] and => To[T]
    */
-  def translateParameterized(from: ClassSymbol, to: ClassSymbol)(implicit ctx: Context): Type = self match {
+  def translateParameterized(from: ClassSymbol, to: ClassSymbol, wildcardArg: Boolean = false)(implicit ctx: Context): Type = self match {
     case self @ ExprType(tp) =>
       self.derivedExprType(tp.translateParameterized(from, to))
     case _ =>
-      if (self.derivesFrom(from)) to.typeRef.appliedTo(self.baseType(from).argInfos)
+      if (self.derivesFrom(from)) {
+        val arg = self.baseType(from).argInfos.head
+        val arg1 = if (wildcardArg) TypeBounds.upper(arg) else arg
+        to.typeRef.appliedTo(arg1)
+      }
       else self
   }
 
@@ -453,7 +457,9 @@ class TypeApplications(val self: Type) extends AnyVal {
   def underlyingIfRepeated(isJava: Boolean)(implicit ctx: Context): Type =
     if (self.isRepeatedParam) {
       val seqClass = if (isJava) defn.ArrayClass else defn.SeqClass
-      translateParameterized(defn.RepeatedParamClass, seqClass)
+      // If `isJava` is set, then we want to turn `RepeatedParam[T]` into `Array[_ <: T]`,
+      // since arrays aren't covariant until after erasure. See `tests/pos/i5140`.
+      translateParameterized(defn.RepeatedParamClass, seqClass, wildcardArg = isJava)
     }
     else self
 

--- a/compiler/src/dotty/tools/dotc/transform/ElimRepeated.scala
+++ b/compiler/src/dotty/tools/dotc/transform/ElimRepeated.scala
@@ -125,6 +125,13 @@ class ElimRepeated extends MiniPhase with InfoTransformer { thisPhase =>
    *  @return  a thicket consisting of `ddef` and a varargs bridge method
    *           which overrides the Java varargs method JM from this phase on
    *           and forwards to `ddef`.
+   *
+   *  A bridge is necessary because the following hold
+   *    - the varargs in `ddef` will change from `RepeatedParam[T]` to `Seq[T]` after this phase
+   *    - _but_ the callers of `ddef` expect its varargs to be changed to `Array[_ <: T]`, since it overrides
+   *      a Java varargs
+   *  The solution is to add a "bridge" method that converts its argument from `Array[_ <: T]` to `Seq[T]` and
+   *  forwards it to `ddef`.
    */
   private def addVarArgsBridge(ddef: DefDef)(implicit ctx: Context): Tree = {
     val original = ddef.symbol.asTerm
@@ -133,7 +140,9 @@ class ElimRepeated extends MiniPhase with InfoTransformer { thisPhase =>
       info = toJavaVarArgs(ddef.symbol.info)).enteredAfter(thisPhase).asTerm
     val bridgeDef = polyDefDef(bridge, trefs => vrefss => {
       val (vrefs :+ varArgRef) :: vrefss1 = vrefss
-      val elemtp = varArgRef.tpe.widen.argTypes.head
+      // Can't call `.argTypes` here because the underlying array type is of the
+      // form `Array[_ <: SomeType]`, so we need `.argInfos` to get the `TypeBounds`.
+      val elemtp = varArgRef.tpe.widen.argInfos.head
       ref(original.termRef)
         .appliedToTypes(trefs)
         .appliedToArgs(vrefs :+ tpd.wrapArray(varArgRef, elemtp))

--- a/tests/pos/i5140/J.java
+++ b/tests/pos/i5140/J.java
@@ -1,0 +1,5 @@
+class Animal {}
+class Dog extends Animal {}
+class J {
+  void foo(Animal... animal) {}
+}

--- a/tests/pos/i5140/S.scala
+++ b/tests/pos/i5140/S.scala
@@ -1,0 +1,9 @@
+class S {
+  val j = new J()
+  val x: Array[Dog] = ???
+  // Check that the java varargs for `foo` gets typed as `Array[_ <: Animal]`.
+  // Otherwise, the call below would fail in -Ycheck:elimRepeated because arrays are invariant before erasure.
+  // This is unsound but allowed.
+  j.foo(x: _*)
+  j.foo(new Dog, new Dog)
+}

